### ISSUE TITLE
Documentation:  Restructure Reference Page

### ISSUE THREAD

**Issue by bundfussr** [pharmaverse/admiral#1105]
### Please select a category the issue is focused on?

Other

### Let us know where something needs a refresh or put your idea here!

The reference page should be restructured:
- Remove “Other Keywords”, move “Timing”, “Sources” to User Utilities, move TTE sources and Metadata to top level
- Add subsection for User Utilities and Developer Utilities/ Various, User Utilities should contain functions called in ADaM scripts, Developer utilities should contain functions called in admiral functions or functions of admiral extension packages. User Utilities should not contain `derive_*` functions.
- Remove "Miscellaneous" section

**Comment by bms63** [comment]
Do you think we could do this one before the release?  @PoojaKumari05 do you think you could do this one as you are familiar with  the pkgdown dramework now?

**Comment by rossfarrugia** [comment]
i wonder if there's any way to separate the more generic functions here vs those targeted and only specific to one single derivation? or might not even be a good idea - but i just wonder how best to get people familiar with the generic functions like the *merged* ones which they can re-use for many study-specific purposes.

**Comment by bundfussr** [comment]
@rossfarrugia , I think it is a good idea. But it is not so easy to implement because it is not clear which functions are generic and which are targeted, e.g., how would you classify functions like `derive_vars_dy()`, `derive_vars_duraton()`, or `derive_var_last_dose_date()`?

**Comment by rossfarrugia** [comment]
yes you're right, might become subjective! i guess the vignettes and ADaM templates are always going to be the key then to helping people learn all our functions, as we cant expect everyone to read through all the reference page as part of their training/onboarding.

**Comment by bundfussr** [comment]
I did not mean that we shouldn't try to restructure the page such that the generic functions are easier to find. I just wanted to say that is is not obvious and we need some time to define clear classes of functions which enables the users to find quickly what they need. One class could be functions which add variables versus functions which create, add, or return observations.

I think the vignettes and templates are helpful if you need to implement the ADaM dataset they are describing. But if I need to implement a study or TA specific variable or parameter, I would not like to read through all the vignettes and templates to find a similar derivation.

**Comment by rossfarrugia** [comment]
yep agree, it definitely still needs a restructure but not how i was suggesting with specific vs generic functions. i like that idea about adding a variable vs adding/editing a record (which would include parameter functions). maybe we just need to try out a few ways to categorise these and throw it past our team for a sanity check. there's probably no perfect way, but i like your angle of thinking to try out

**Comment by bms63** [comment]
How about categorizing by how many times they appear across typical specs.  For example, ADY appears across many different datasets, but R2ANRLO only appears in ADLB.  

So...specific versus generic would still be helpful.  Maybe instead of generic there would a tag like `common-add vars` and `common-add obs` and then a `specific-add vars` and `specific-add vars`

`common-add vars` - Variable derivation appears in 2 or more datasets across typical specs
`common-add obs` - Observation derivation appears in 2 or more datasets across typical specs
`specific-add vars` - Variable derivation appears in 2 or less datasets across typical specs
`specific-add obs` - Observation derivation appears in 2 or less datasets across typical specs

Not beholden to my ingenious naming scheme.

**Comment by rossfarrugia** [comment]
hmm... you'd lose the benefit then of having them alphabetised which is actually really useful when you know the function name you're looking for and just want to scroll down to it. i'm personally more in favour of Stefan's proposal of categorising by similar behaviour (and then alphabetising). probably a good discussion for our call tomorrow!

**Comment by bms63** [comment]
I end up using the search bar or doing a control+f to search for finding functions, but yes we would lose the alphabetised reference if we go my way.  

However, I'm a bit more interested in this as I have yet to use `derive_derive_param` (first function on the current list), but have used `derive_var_ady`, `derive_vars_merged_dtm` across multiple datasets.  I would imagine these would be typical derivations that new comers would like to see.

Yes!  Good topic for meeting tomorrow!!

**Comment by bundfussr** [comment]
Using `@family` and `meta.R` looks promising.

**Comment by bms63** [comment]
I've been playing around with the families.  Here are some examples that could appear in our documentation.  You can mess with the meta.R file to make the text a little nicer on the eyes - Hence the :

![image](https://user-images.githubusercontent.com/10111024/176277407-23605c4a-fa5b-4c58-bc01-12f4de6a07b5.png)
![image](https://user-images.githubusercontent.com/10111024/176277545-bded1f1d-bb26-45fc-b016-89a3224759c3.png)
![image](https://user-images.githubusercontent.com/10111024/176277736-624eedea-54d5-44e0-98e1-4f1189f33ef7.png)
![image](https://user-images.githubusercontent.com/10111024/176277793-36e4047a-1e8e-410a-a7d4-a8b173c6ca49.png)

You can see I have 4 families that I have created:
* Date/Time Functions
* Common BDS Functions
* Utilities for Dataset Checking
* Utilities for Filtering Observations

In no way are these set in stone!  

Other family candidates could be:

* Utilities for Observation
* Utilities for Formatting
* Utilities for Units
* Common Flag Variable Derivations

Suggestion for others?

**Comment by bundfussr** [comment]
We should also think about if and how deprecated functions are listed (see https://github.com/pharmaverse/admiral/issues/1262).

**Comment by bms63** [comment]
Hi all,

I've been coming up with more families.  I'm just flailing in the wind here - so don't hesitate to be critical of my themes.

Thinking that we should split out families for compute and derivations.  For example, compute functions for common bds would be grouped together and derive functions for common bds would be grouped together.  This way a user looking at the function `derive_var_analysis_ratio` would see other bds findings function grouped in the documetation like `derive_param_qtc` and `derive_param_wbc_abs`.

![image](https://user-images.githubusercontent.com/10111024/176557093-704d40ff-41a2-4630-860a-38ea293a4db4.png)

**Comment by rossfarrugia** [comment]
Definitely agree to keep compute and derive functions separate in the list @bms63 as its important newcomers understand the different behaviour. we had this question a lot in the external testing.

**Comment by bundfussr** [comment]
All these families make sense to me. It seems that the families are combinations of existing or not yet existing keywords. For example, the `com_date_time` family includes all functions with keywords `computation` and `timing`.

I am a bit concerned regarding the burden for keeping the families and keywords consistent and up to date.

I wonder if we could automate it partially. For example we could add the `@keywords` tag to the header but not the `@family` tag. If the documentation is rendered, a "Related Functions" section is added. It lists all functions with the keywords from the `@keywords` tag. The keywords for listing the functions could be modified by the user. For example, if we have `@keywords computation timing`, the user could deactivate the `computation` keyword to display all timing related functions.

With this approach we would need to maintain the keywords only. And the users could decide which famlies (= combination of keywords) are useful for them.

As a user and developer I would like this approach but I am not sure if it is technically possible.

What do you think?

**Comment by bms63** [comment]
umpf... @bundfussr sounds amazing but way beyond my paltry skills!  

So keywords are for helping with organizing the reference section.  I'm actually learning towards just one keyword per function.  This way we can do `starts_with(computation)` and it only appears once.  However, I think keywords also need an update as it would be nice to separate the reference section with one keywords so `starts_with(computation)` would be `starts_with(com_date_time)` and then the family would be `com_date_time`.  So a developer only has to use the same keyword and family for their function.  

This would mean that compute functions dealing with date/time would be organized nicely on the reference page and the individual documentation would link out to similar functions within the family.  

The rate of adding new admiral functions has decreased a lot.  I think have some way to extract the keywords and families and have a high-level overview will be really helpful.  I have some code to do this now...so hopefully we stay on top of this.  Perhaps add it to the CRAN checklist.

**Comment by bms63** [comment]
Here is deprecated section of the reference page.  These will have to have keywords deprecated and will need to be part of our deprecation guidance .  I will update the programming strategy.

**Comment by bms63** [comment]
Updated Deprecation Guidance in programming strategy
 
![image](https://user-images.githubusercontent.com/10111024/176693896-462833aa-123e-4f53-8d69-1d8b1bea2b30.png)
![image](https://user-images.githubusercontent.com/10111024/176695261-6e5e743c-ddc1-4182-b846-19b012fcd42e.png)

**Comment by bundfussr** [comment]
@bms63 , I am not sure what you mean by "just one keyword per function". Do you want to list all derivations in the "Derivations" section and nowhere else? For example, `derive_param_bmi()` would be listed in "Derivations" but not in "ADVS"?

**Comment by bms63** [comment]
I'm leaning towards that so it  just make the reference page not so verbose.  But im open to suggestions.

**Comment by bms63** [comment]
Here is example of date/time derivations and computations listed directly on reference page as seen on first image.  Once you you click on, for example `derive_vars_dt`, in the **See Also** section you can see similar functions for date/time derivations. 

![image](https://user-images.githubusercontent.com/10111024/176723570-dbc8dfac-b027-4ee7-89b6-1bad37e3f9ec.png)

![image](https://user-images.githubusercontent.com/10111024/176724049-bb3b79d5-d8ba-4abc-b84d-8a689faf558f.png)

**Comment by bms63** [comment]
@morrishj  I'm working in this branch if you want to see what I've been tweaking https://github.com/pharmaverse/admiral/tree/1105_restructure_docs

**Comment by bms63** [comment]
Okay!  Phew...this is tedious.  Definitely earning my ration of alcohol this weekend.  

Got some more themes going on here on the right side of the Reference page.    

The `@keywords` make it easy to group functions on the reference page.  I think we should only do one keyword as having it listed multiple times on the same page is confusing - at least for me.  If a user goes to the function docs and doesn't like what they see, then in the `@family` will list out additional functions for them.

![image](https://user-images.githubusercontent.com/10111024/176806229-ea60f9aa-699b-47b1-987f-f37000c43c3f.png)

The `@family` make it easy to list out all the functions grouped on the references page that a user might like on a specific function page.  

I'm happy with the `Functions moved to another pharmaverse package` section :)  

![image](https://user-images.githubusercontent.com/10111024/176806620-02a4abf3-b9a5-43ef-ad5c-9609fea24525.png)

We can also still use `@seealso` to help with a user that might need a very specific function reference

**Comment by bundfussr** [comment]
@bms63 , I am not sure if the new structure works. I think the main issue is that the new sections are not mutually exclusive but each function is listed only once. For example, `derivat_var_trtdur()` is listed in "Derivations for Date/Time Variables", which makes sense. However, it is also a standard ADSL derivation but it is not listed in "Derivations for ADSL". So users may miss it if they look in "Derivations for ADSL" only.

**Comment by bms63** [comment]
HI all.  I managed to fork admiral and republish the website on my own repository.  Hoping this can help to spur discussion on rework of the reference page.

Lots of mis-spellings and unfurnished pieces but nice to have something to look at and try out.  It is actually kind of cool as I can work in admiral branch `1105_restrucutre_docs` and then fetch my updates in my repo and republish the site from my repo.  This way people don't have to run things locally or I have to post screen shots of proposals.    

https://bms63.github.io/admiral/main/reference/index.html

**Comment by rossfarrugia** [comment]
Thanks @bms63 - helps to have something to play around with and explore. I like the new direction, but here would be my suggested simpler wordings with any new categories in bold. Generally I feel like less is more with sub-categories as the more we add the harder it will be to maintain and you'd have difficult decisions like I could put this function in multiple (to Stefan's point above)

- Derivations for adding variables
  - All ADaMs _(would include Date/Time)_
  - ADSL-specific
  - BDS-specific
  - OCCDS-specific 
- **Derivations for adding parameters**
  - BDS-specific
  - TTE-specific 
- Advanced Functions
  - Higher Order
  - Metadata
  - Other
- Computation Functions for Vectors
  - All ADaMs 
  - BDS-specific
- Utilities _(I would remove all the sub-categories here)_
- Example Datasets
- Relocated/Depracated

I didn't correct any spellings above as trust you'll get these later :) happy to be over-ruled on the above but having played around with your reference page, I think this might be simpler!

**Comment by bms63** [comment]
Not purposeful.  will update

**Comment by bms63** [comment]
> Really looks nice.
> 
> Some comments and initial thoughts:
> 
> * advanced functions - should we add a bit more detail what this means? I don't want to add a lot of documentation in addition, but I wonder if it is clear for new users what "advanced functions" could mean. Basically one or two bullets that summarize what advanced functions purpose is and why these been created? (i.e. not ADaM dataset specific, higher order functions functions to slide, restrict datasets to apply calculations on selected data onyl ...) .

I think this is a good idea.  I would like to link out to vignettes that deal with these functions as they are slightly mysterious to me as well!!

> * I wonder if the computation functions for BDS should be in the BDS section? I know they differ from the create a param/variable in BDS but somehow it might be easier if the users see all BDS functions in one place?
> 

Users have reported confusion around these derivation and compute functions being together in the past and not knowing which one to use.  I think keeping them separate should be the way forward.  

> Relocated functions - could we add where these been relocated to? Or is that too much maintenance?
> 

I put a blurb in about assertion and other functions moving to `admiraldev`.  Hopefully can link the admiraldev site when that comes online.  I think this will be a wee bit of hassle for us to maintain but a really nice to have as we move things around between our different packages.  Also helps us developers keep a eye on what is in what package!

> Deprecated function - could we already add what they will be replaced with? Or is that too much maintenance?
> 

The deprecated function has that type of information in it when you run it.  I think this might be a bit of a hassle to write into the website - not entirely sure how that would be done.

> Thanks a lot @bms63 for re-structuring this. I still remember I had this as an first issue to think about a structure/categories before your time :) Now it seems to come into place quite nicely.

:)

**Comment by bms63** [comment]
Hi all,

So I'm making an editorial decision here.  

I want the `@keywords` and `@family` to be **identical** and for there only to be **one** entry for both.   This makes development and maintenance way simpler and makes it so our developers don't use multiple keywords for a function that sometimes don't make sense (or make up their own!)

I've updated the https://bms63.github.io/admiral/main/articles/programming_strategy.html#categorization-of-functions to reflect my decision on this issue.  

I've also listed out the only keywords and families available to developers in this section.  We can add to this list as needed or clarify in the description.  I might need some word-smithing help here.

So...

A developer making a new function needs to decide what keyword/family their functions fits into OR ask for a new keyword/family to be created.  Once they add this keyword/family (from the proscribed list), then their function will be added into both the relevant Reference Section and the See Also section of related family of functions.

**Comment by bms63** [comment]
Also,

I extracted all the keyword/family combos and placed in Extraction July-11 if if you are interested in the keyword/family combos for each function.

https://exbpbox.box.com/s/aj6t1c2tokoxs0g2a7in1yu7ioy136gk

![image](https://user-images.githubusercontent.com/10111024/178334523-ac26f530-5267-48f6-9aa0-43b91a07c4a6.png)

**Comment by bms63** [comment]
See #688
